core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -415,9 +420,54 @@ private[spark] class TaskSchedulerImpl(
             launchedAnyTask |= launchedTaskAtCurrentMaxLocality
           } while (launchedTaskAtCurrentMaxLocality)
         }
+
         if (!launchedAnyTask) {
-          taskSet.abortIfCompletelyBlacklisted(hostToExecutors)
+          taskSet.getCompletelyBlacklistedTaskIfAny(hostToExecutors) match {
+            case Some(taskIndex) => // Returns the taskIndex which was unschedulable
+
+              // If the taskSet is unschedulable we try to find an existing idle blacklisted
+              // executor. If we cannot find one, we abort immediately. Else we kill the idle
+              // executor and kick off an abortTimer which if it doesn't schedule a task within the
+              // the timeout will abort the taskSet if we were unable to schedule any task from the
+              // taskSet.
+              // Note 1: We keep track of schedulability on a per taskSet basis rather than on a per
+              // task basis.
+              // Note 2: The taskSet can still be aborted when there are more than one idle
+              // blacklisted executors and dynamic allocation is on. This can happen when a killed
+              // idle executor isn't replaced in time by ExecutorAllocationManager as it relies on
+              // pending tasks and doesn't kill executors on idle timeouts, resulting in the abort
+              // timer to expire and abort the taskSet.
+              executorIdToRunningTaskIds.find(x => !isExecutorBusy(x._1)) match {
+                case Some ((executorId, _)) =>
+                  if (!unschedulableTaskSetToExpiryTime.contains(taskSet)) {
+                    blacklistTrackerOpt.foreach(blt => blt.killBlacklistedIdleExecutor(executorId))
+
+                    val timeout = conf.get(config.UNSCHEDULABLE_TASKSET_TIMEOUT) * 1000
+                    unschedulableTaskSetToExpiryTime(taskSet) = clock.getTimeMillis() + timeout
+                    logInfo(s"Waiting for $timeout ms for completely "
+                      + s"blacklisted task to be schedulable again before aborting $taskSet.")
+                    abortTimer.schedule(
+                      createUnschedulableTaskSetAbortTimer(taskSet, taskIndex), timeout)
+                  }
+                case _ => // Abort Immediately
+                  logInfo("Cannot schedule any task because of complete blacklisting. No idle" +
+                    s" executors can be found to kill. Aborting $taskSet." )
+                  taskSet.abortSinceCompletelyBlacklisted(taskIndex)
+              }
+            case _ => // Do nothing if no tasks completely blacklisted.
+          }
+        } else {
+          // We want to defer killing any taskSets as long as we have a non blacklisted executor
+          // which can be used to schedule a task from any active taskSets. This ensures that the
+          // job can make progress and if we encounter a flawed taskSet it will eventually either
+          // fail or abort due to being completely blacklisted.
```

Comment:
I think you should say here that you may have a job wait indefinitely, if its effectively blacklisted the entire cluster, but other jobs keep coming in and keeping resources occupied so the cluster stays busy.  So its not really accurate to say that it will be aborted eventually, we are actually *not* guaranteeing that (if I understood things correctly).

Since its folded now lemme reference the prior discussion on this: https://github.com/apache/spark/pull/22288#discussion_r226477452

>> Want to make sure I understand this part, and why you aren't only clearing the timer for the taskset you just scheduled a task for. If you have multiple tasksets running simultaneously, one is making progress but the other is totally blacklisted, I guess you do not want to kill anything, because that would mess with the taskset that is working correctly? Instead you'll just let the taskset which is totally blacklisted eventually fail from the timeout? I guess that makes sense, because if one taskset is progressing, it means the failing taskset probably is probably flawed, not the executors.
>>
>>If that's right, would be good to include something along those lines in the comment (personally I don't find a comment about how its related to the timer that useful, that's obvious from the code).
>
 >dhruve 7 days ago  Contributor
>That is correct. It also covers other scenario that @tgravescs originally pointed out.
>
>Lets say if you have multiple taskSets running which are completely blacklisted. If you were able to get an executor, you would just clear the timer for that specific taskSet. Now due to resource constraint, if you weren't able to obtain another executor within the timeout for the other taskSet, you would abort the other taskSet when you could actually wait for it to be scheduled on the newly obtained executor.
>
> So clearing the timer for all the taskSets ensures that currently we aren't in a completely blacklisted state and should try to run to completion. However if the taskset itself is flawed, we would eventually fail. This could result in wasted effort, but we don't have a way to determine that yet, so this should be okay.

Reply:
Your understanding is correct. I will update the comment.

Reply:
@tgravescs since we've been back and forth on the discussion of the cases here, just want to make sure you're aware of the possibility for waiting indefinitely here.

Reply:
Thanks for pointing this out, but if I'm reading the discussion properly, I don't think you will actually wait indefinitely.  Eventually you will either abort immediately or you should fail due to max number of task failures.  Let me know if I'm missing something from the scenario.

Lets say you have taskset1 that is blacklisted on all nodes (lets say we have 3). 3 cases can happen at this point:

- taskset 2 hasn't started, so it tries to kill an executor and starts the timer.  
- taskset 2 has started, if its running on all nodes then we abort immediately because no executors to kill to kill
- taskset 2 has started but its not running on all blacklisted nodes, then we will kill an executor

At this point lets say we didn't abort so we killed an executor.  Taskset 1 will get a chance to run on the new executor and either work or have a task failure.  If it has a task failure and it gets blacklisted, we go back into the case above.  But the # of task failures gets one closer.

so it seems like eventually you would either abort immediately if there aren't any executors to kill or you would eventually fail with max number of task attempts.

Reply:
Here's the scenario I'm worried about:

1) taskset1 and taskset2 are both running currently.  taskset1 has enough failures to get blacklisted everywhere.
2) there is an idle executor, even though taskset2 is running (eg. the executor that is available doesn't meet the locality preferences of taskset2).  So abort timer is started.
3) the idle executor is killed, and you get a new one.
4) just by luck, taskset2 gets a hold of the new idle executor (eg. the executor is on a node blacklisted by taskset1, or taskset2 just has a higher priority).  abort timer is cleared
5) taskset2 finishes, but meanwhile taskset3 has been launched, and it uses the idle executor.   etc. for taskSetN, so you keep launching tasks, abort timer gets cleared, but nothing even gets scheduled on taskset1.

admittedly this would not be the normal scenario -- you'll need more tasksets to keep coming, and you need tight enough resource constraints that taskset1 never get a hold of anything, even the new one.

Reply:
ok, yeah it seems like it would have to be very timing dependent that taskset1 never got a chance for that executor, really that would just be a normal indefinite postponement problem in the scheduler regardless of blacklisting.   I don't think with fifo its a problem as first taskset should always be first. With Fair scheduler perhaps it could but probably depends on much more specific scenario.   

I guess I'm ok with this if you are.